File: forms/outpatient-outcome/back-end-with-loco/src/engine/operational_domain.rs
//! Operational domain module.

use super::types::{AssessmentData, DomainGrade, FiredRule};

/// Operational result.
pub struct OperationalResult {
    /// Grade.
    pub grade: DomainGrade,
    /// Rules.
    pub rules: Vec<FiredRule>,
}

/// Grade the Operational domain from NHS Attendance Outcome + wait-time vs target.
pub fn grade_operational(data: &AssessmentData) -> OperationalResult {
    let outcome = data.operational_efficiency.nhs_attendance_outcome.as_str();
    let wait = data.operational_efficiency.wait_time_days;
    let target = data.operational_efficiency.service_target_days;
    let mut rules: Vec<FiredRule> = Vec::new();

    if outcome.is_empty() {
        return OperationalResult { grade: String::new(), rules };
    }

    let dna_or_provider = outcome == "patient_dna" || outcome == "provider_cancelled";
    let patient_cancelled = outcome == "patient_cancelled";
    let attended = matches!(
        outcome,
        "attended_discharged" | "attended_follow_up" | "attended_pifu" | "attended_onward_referral"
    );

    if dna_or_provider {
        let desc = if outcome == "patient_dna" {
            "Appointment outcome: Did Not Attend".to_string()
        } else {
            "Appointment outcome: Provider Cancelled".to_string()
        };
        rules.push(FiredRule {
            id: "OPS-004".to_string(),
            domain: "Operational".to_string(),
            description: desc,
            grade: "E".to_string(),
        });
        return OperationalResult { grade: "E".to_string(), rules };
    }

    if patient_cancelled {
        rules.push(FiredRule {
            id: "OPS-003".to_string(),
            domain: "Operational".to_string(),
            description: "Appointment cancelled or rebooked by patient".to_string(),
            grade: "D".to_string(),
        });
        return OperationalResult { grade: "D".to_string(), rules };
    }

    if attended {
        if let (Some(w), Some(t)) = (wait, target) {
            if t > 0 {
                let ratio = w as f64 / t as f64;
                if ratio <= 1.0 {
                    rules.push(FiredRule {
                        id: "OPS-001".to_string(),
                        domain: "Operational".to_string(),
                        description: format!(
                            "Attended within target ({w}d ≤ {t}d target)"
                        ),
                        grade: "A".to_string(),
                    });
                    return OperationalResult { grade: "A".to_string(), rules };
                } else if ratio <= 1.5 {
                    rules.push(FiredRule {
                        id: "OPS-002A".to_string(),
                        domain: "Operational".to_string(),
                        description: format!(
                            "Attended within 1.5× target ({w}d vs {t}d target)"
                        ),
                        grade: "B".to_string(),
                    });
                    return OperationalResult { grade: "B".to_string(), rules };
                } else {
                    rules.push(FiredRule {
                        id: "OPS-002B".to_string(),
                        domain: "Operational".to_string(),
                        description: format!(
                            "Attended but wait exceeded 1.5× target ({w}d vs {t}d target)"
                        ),
                        grade: "C".to_string(),
                    });
                    return OperationalResult { grade: "C".to_string(), rules };
                }
            }
        }
        // Attended but no wait data: default to B with a data-quality note.
        rules.push(FiredRule {
            id: "OPS-001B".to_string(),
            domain: "Operational".to_string(),
            description: "Attended; wait-time or target not recorded (assumed within range)"
                .to_string(),
            grade: "B".to_string(),
        });
        return OperationalResult { grade: "B".to_string(), rules };
    }

    OperationalResult { grade: String::new(), rules }
}

File: forms/outpatient-outcome/front-end-with-svelte/src/lib/engine/operational-domain.ts
import type { AssessmentData, DomainGrade, FiredRule } from './types';

/**
 * Grade the Operational domain from NHS Attendance Outcome + wait-time vs target.
 *
 * Grade rules:
 * A = Attended AND wait ≤ service target
 * B = Attended AND wait ≤ 1.5 × service target
 * C = Attended AND wait > 1.5 × service target
 * D = Patient cancelled or rebooked (patient_cancelled)
 * E = DNA (patient_dna) OR provider-cancelled (provider_cancelled)
 * '' = Missing attendance outcome code
 *
 * "Attended" outcomes: attended_discharged, attended_follow_up, attended_pifu,
 * attended_onward_referral
 */
export function gradeOperational(data: AssessmentData): { grade: DomainGrade; rules: FiredRule[] } {
	const { nhsAttendanceOutcome, waitTimeDays, serviceTargetDays } = data.operationalEfficiency;
	const rules: FiredRule[] = [];

	if (!nhsAttendanceOutcome) {
		return { grade: '', rules };
	}

	const dnaOrProviderCancelled =
		nhsAttendanceOutcome === 'patient_dna' || nhsAttendanceOutcome === 'provider_cancelled';

	const patientCancelled = nhsAttendanceOutcome === 'patient_cancelled';

	const attended = [
		'attended_discharged',
		'attended_follow_up',
		'attended_pifu',
		'attended_onward_referral'
	].includes(nhsAttendanceOutcome);

	if (dnaOrProviderCancelled) {
		rules.push({
			id: 'OPS-004',
			domain: 'Operational',
			description: `Appointment outcome: ${nhsAttendanceOutcome === 'patient_dna' ? 'Did Not Attend' : 'Provider Cancelled'}`,
			grade: 'E'
		});
		return { grade: 'E', rules };
	}

	if (patientCancelled) {
		rules.push({
			id: 'OPS-003',
			domain: 'Operational',
			description: 'Appointment cancelled or rebooked by patient',
			grade: 'D'
		});
		return { grade: 'D', rules };
	}

	if (attended) {
		// Check wait time vs target
		if (waitTimeDays !== null && serviceTargetDays !== null && serviceTargetDays > 0) {
			const ratio = waitTimeDays / serviceTargetDays;
			if (ratio <= 1) {
				rules.push({
					id: 'OPS-001',
					domain: 'Operational',
					description: `Attended within target (${waitTimeDays}d ≤ ${serviceTargetDays}d target)`,
					grade: 'A'
				});
				return { grade: 'A', rules };
			} else if (ratio <= 1.5) {
				rules.push({
					id: 'OPS-002A',
					domain: 'Operational',
					description: `Attended within 1.5× target (${waitTimeDays}d vs ${serviceTargetDays}d target)`,
					grade: 'B'
				});
				return { grade: 'B', rules };
			} else {
				rules.push({
					id: 'OPS-002B',
					domain: 'Operational',
					description: `Attended but wait exceeded 1.5× target (${waitTimeDays}d vs ${serviceTargetDays}d target)`,
					grade: 'C'
				});
				return { grade: 'C', rules };
			}
		} else {
			// Attended but no wait-time data → default to B (attended, data quality flag)
			rules.push({
				id: 'OPS-001B',
				domain: 'Operational',
				description: 'Attended; wait-time or target not recorded (assumed within range)',
				grade: 'B'
			});
			return { grade: 'B', rules };
		}
	}

	return { grade: '', rules };
}
